### PR TITLE
Implement Trylock

### PR DESCRIPTION
**Pull request by soypat**
~I also fixed something which _seems_ like a bug... The unlock method was not resetting the value of the `locked` field. I suspect there may be more than meets the eye and that this is not a bug...~

I now understand what the code is doing, removed the bug I introduced via force push :facepalm:

### CODE REVIEW

File: src/sync/mutex.go
Comment (dgryski) [inline src/sync/mutex.go]: It wouldn't surprise me if we needed calls to volatile loads here: https://github.com/tinygo-org/tinygo/blob/release/src/runtime/volatile/volatile.go

Comment (soypat) [reply src/sync/mutex.go]: :thinking: Wouldn't a memory barrier be better suited for this use case? Even then, tinygo is single threaded so we should have consistent ordering?

Comment (dgryski) [reply src/sync/mutex.go]: I'm more concerned about LLVM removing loads/unloads it doesn't think are important.  Yes, TInyGo is single threaded so memory barriers are less interesting.

Comment (soypat) [reply src/sync/mutex.go]: I've added volatile loads/stores as requested!

**Comment by deadprogram** [comment]
Looking at this code, seems like it should be OK to squash/merge.

@soypat do you want to squash the commits yourself?
@aykevl or anyone else any other feedback before merge?

**Review by aykevl — APPROVED** [review]
Looks good to me, though if we're going to care about atomicity we should really use the sync/atomic package instead. Volatile has slightly different semantics (though for most microcontrollers it doesn't really matter).

Rebased on the dev branch, ready to merge once CI passes.

**Pull request merged**
